### ISSUE TITLE
Workaround for gnu compiler bug (7.2.0 and later): assigning to character array via associate

### ISSUE THREAD

**Issue by billsacks** [ESCOMP/ctsm#68]
**Bill Sacks < sacks@ucar.edu > - 2017-09-20 15:38:17 -0600**
**Bugzilla Id:** [2513](http://bugz.cgd.ucar.edu/show_bug.cgi?id=2513)
**Bugzilla CC:**  andre@ucar.edu,  jedwards@ucar.edu,  rfisher@ucar.edu, 

This bug in recent versions of gfortran (affecting 7.2.0, 8.0 and possibly other versions):

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82121

means that we get compilation errors like this:

/Users/jedwards/cesm/cesm2_0_alpha/components/clm/src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90:425:6:

       decomp_pool_name_restart(i_litr1) = 'litr1'

      1

Error: Unclassifiable statement at (1)

/Users/jedwards/cesm/cesm2_0_alpha/components/clm/src/soilbiogeochem/SoilBiogeochemDecompCascadeBGCMod.F90:426:6:

       decomp_pool_name_history(i_litr1) = 'LITR1'

      1

A workaround would be to set these variables directly rather than via associate statements. We could try doing a find & replace of decomp_pool_name  with decomp_cascade_con%decomp_pool_name. I'm not sure if this issue appears in other places, too.

**Comment by billsacks** [comment]
**Erik Kluzek < erik@ucar.edu > - 2017-09-20 16:09:06 -0600**

I did some checking, and character data is rare enough, especially in associate statements that I think the only other place this occurs is: 

components/clm/src/soilbiogeochem/SoilBiogeochemDecompCascadeCNMod.F90

It also looks like this workaround is reasonably straightforward. Here it is for the BGC version...

Index: SoilBiogeochemDecompCascadeBGCMod.F90
===================================================================
--- SoilBiogeochemDecompCascadeBGCMod.F90	(revision 86700)
+++ SoilBiogeochemDecompCascadeBGCMod.F90	(working copy)
@@ -361,10 +361,6 @@
          cascade_donor_pool             => decomp_cascade_con%cascade_donor_pool                 , & ! Output: [integer           (:)     ]  which pool is C taken from for a given decomposition step 
          cascade_receiver_pool          => decomp_cascade_con%cascade_receiver_pool              , & ! Output: [integer           (:)     ]  which pool is C added to for a given decomposition step   
          floating_cn_ratio_decomp_pools => decomp_cascade_con%floating_cn_ratio_decomp_pools     , & ! Output: [logical           (:)     ]  TRUE => pool has fixed C:N ratio                          
-         decomp_pool_name_restart       => decomp_cascade_con%decomp_pool_name_restart           , & ! Output: [character(len=8)  (:)     ]  name of pool for restart files                   
-         decomp_pool_name_history       => decomp_cascade_con%decomp_pool_name_history           , & ! Output: [character(len=8)  (:)     ]  name of pool for history files                   
-         decomp_pool_name_long          => decomp_cascade_con%decomp_pool_name_long              , & ! Output: [character(len=20) (:)     ]  name of pool for netcdf long names              
-         decomp_pool_name_short         => decomp_cascade_con%decomp_pool_name_short             , & ! Output: [character(len=8)  (:)     ]  name of pool for netcdf short names              
          is_litter                      => decomp_cascade_con%is_litter                          , & ! Output: [logical           (:)     ]  TRUE => pool is a litter pool                             
          is_soil                        => decomp_cascade_con%is_soil                            , & ! Output: [logical           (:)     ]  TRUE => pool is a soil pool                               
          is_cwd                         => decomp_cascade_con%is_cwd                             , & ! Output: [logical           (:)     ]  TRUE => pool is a cwd pool                                
@@ -422,10 +418,10 @@
 
       !-------------------  list of pools and their attributes  ------------
       floating_cn_ratio_decomp_pools(i_litr1) = .true.
-      decomp_pool_name_restart(i_litr1) = 'litr1'
-      decomp_pool_name_history(i_litr1) = 'LITR1'
-      decomp_pool_name_long(i_litr1) = 'litter 1'
-      decomp_pool_name_short(i_litr1) = 'L1'
+      decomp_cascade_con%decomp_pool_name_restart(i_litr1) = 'litr1'
+      decomp_cascade_con%decomp_pool_name_history(i_litr1) = 'LITR1'
+      decomp_cascade_con%decomp_pool_name_long(i_litr1) = 'litter 1'
+      decomp_cascade_con%decomp_pool_name_short(i_litr1) = 'L1'
       is_litter(i_litr1) = .true.
       is_soil(i_litr1) = .false.
       is_cwd(i_litr1) = .false.
@@ -436,10 +432,10 @@
       is_lignin(i_litr1) = .false.
 
       floating_cn_ratio_decomp_pools(i_litr2) = .true.
-      decomp_pool_name_restart(i_litr2) = 'litr2'
-      decomp_pool_name_history(i_litr2) = 'LITR2'
-      decomp_pool_name_long(i_litr2) = 'litter 2'
-      decomp_pool_name_short(i_litr2) = 'L2'
+      decomp_cascade_con%decomp_pool_name_restart(i_litr2) = 'litr2'
+      decomp_cascade_con%decomp_pool_name_history(i_litr2) = 'LITR2'
+      decomp_cascade_con%decomp_pool_name_long(i_litr2) = 'litter 2'
+      decomp_cascade_con%decomp_pool_name_short(i_litr2) = 'L2'
       is_litter(i_litr2) = .true.
       is_soil(i_litr2) = .false.
       is_cwd(i_litr2) = .false.
@@ -450,10 +446,10 @@
       is_lignin(i_litr2) = .false.
 
       floating_cn_ratio_decomp_pools(i_litr3) = .true.
-      decomp_pool_name_restart(i_litr3) = 'litr3'
-      decomp_pool_name_history(i_litr3) = 'LITR3'
-      decomp_pool_name_long(i_litr3) = 'litter 3'
-      decomp_pool_name_short(i_litr3) = 'L3'
+      decomp_cascade_con%decomp_pool_name_restart(i_litr3) = 'litr3'
+      decomp_cascade_con%decomp_pool_name_history(i_litr3) = 'LITR3'
+      decomp_cascade_con%decomp_pool_name_long(i_litr3) = 'litter 3'
+      decomp_cascade_con%decomp_pool_name_short(i_litr3) = 'L3'
       is_litter(i_litr3) = .true.
       is_soil(i_litr3) = .false.
       is_cwd(i_litr3) = .false.
@@ -466,10 +462,10 @@
       if (.not. use_fates) then
          ! CWD
          floating_cn_ratio_decomp_pools(i_cwd) = .true.
-         decomp_pool_name_restart(i_cwd) = 'cwd'
-         decomp_pool_name_history(i_cwd) = 'CWD'
-         decomp_pool_name_long(i_cwd) = 'coarse woody debris'
-         decomp_pool_name_short(i_cwd) = 'CWD'
+         decomp_cascade_con%decomp_pool_name_restart(i_cwd) = 'cwd'
+         decomp_cascade_con%decomp_pool_name_history(i_cwd) = 'CWD'
+         decomp_cascade_con%decomp_pool_name_long(i_cwd) = 'coarse woody debris'
+         decomp_cascade_con%decomp_pool_name_short(i_cwd) = 'CWD'
          is_litter(i_cwd) = .false.
          is_soil(i_cwd) = .false.
          is_cwd(i_cwd) = .true.
@@ -486,10 +482,10 @@
          i_soil1 = 4
       endif
       floating_cn_ratio_decomp_pools(i_soil1) = .false.
-      decomp_pool_name_restart(i_soil1) = 'soil1'
-      decomp_pool_name_history(i_soil1) = 'SOIL1'
-      decomp_pool_name_long(i_soil1) = 'soil 1'
-      decomp_pool_name_short(i_soil1) = 'S1'
+      decomp_cascade_con%decomp_pool_name_restart(i_soil1) = 'soil1'
+      decomp_cascade_con%decomp_pool_name_history(i_soil1) = 'SOIL1'
+      decomp_cascade_con%decomp_pool_name_long(i_soil1) = 'soil 1'
+      decomp_cascade_con%decomp_pool_name_short(i_soil1) = 'S1'
       is_litter(i_soil1) = .false.
       is_soil(i_soil1) = .true.
       is_cwd(i_soil1) = .false.
@@ -505,10 +501,10 @@
          i_soil2 = 5
       endif
       floating_cn_ratio_decomp_pools(i_soil2) = .false.
-      decomp_pool_name_restart(i_soil2) = 'soil2'
-      decomp_pool_name_history(i_soil2) = 'SOIL2'
-      decomp_pool_name_long(i_soil2) = 'soil 2'
-      decomp_pool_name_short(i_soil2) = 'S2'
+      decomp_cascade_con%decomp_pool_name_restart(i_soil2) = 'soil2'
+      decomp_cascade_con%decomp_pool_name_history(i_soil2) = 'SOIL2'
+      decomp_cascade_con%decomp_pool_name_long(i_soil2) = 'soil 2'
+      decomp_cascade_con%decomp_pool_name_short(i_soil2) = 'S2'
       is_litter(i_soil2) = .false.
       is_soil(i_soil2) = .true.
       is_cwd(i_soil2) = .false.
@@ -524,10 +520,10 @@
          i_soil3 = 6
       endif
       floating_cn_ratio_decomp_pools(i_soil3) = .false.
-      decomp_pool_name_restart(i_soil3) = 'soil3'
-      decomp_pool_name_history(i_soil3) = 'SOIL3'
-      decomp_pool_name_long(i_soil3) = 'soil 3'
-      decomp_pool_name_short(i_soil3) = 'S3'
+      decomp_cascade_con%decomp_pool_name_restart(i_soil3) = 'soil3'
+      decomp_cascade_con%decomp_pool_name_history(i_soil3) = 'SOIL3'
+      decomp_cascade_con%decomp_pool_name_long(i_soil3) = 'soil 3'
+      decomp_cascade_con%decomp_pool_name_short(i_soil3) = 'S3'
       is_litter(i_soil3) = .false.
       is_soil(i_soil3) = .true.
       is_cwd(i_soil3) = .false.

**Comment by billsacks** [comment]
**Bill Sacks < sacks@ucar.edu > - 2017-10-17 16:09:43 -0600**

Changing this to P1 because we feel it's important to resolve this for the release.

**Comment by ekluzek** [comment]
Fixed in clm5.0.dev0004

**Comment by ekluzek** [comment]
Reopening. Because, there are some lines that still need to be addressed, such as these...

 ```
     !----------------  list of transitions and their time-independent coefficients  ---------------!
      i_l1s1 = 1
      cascade_step_name(i_l1s1) = 'L1S1'

```

**Comment by ekluzek** [comment]
The lines that need to change involved cascade_step_name from line 561 to 625. The CN version has the same issue as well.

**Comment by ekluzek** [comment]
Fixed in PR #377 came in with clm5.0.dev010
